core/src/main/java/com/linecorp/armeria/server/HttpHeaderUtil.java

```diff
@@ -0,0 +1,186 @@
+import com.linecorp.armeria.common.HttpHeaders;
+
+import io.netty.util.AsciiString;
+import io.netty.util.NetUtil;
+
+/**
+ * A utility class which provides useful methods for handling HTTP headers.
+ */
+final class HttpHeaderUtil {
+    private static final Logger logger = LoggerFactory.getLogger(HttpHeaderUtil.class);
+
+    private static final Splitter CSV_SPLITTER = Splitter.on(',').omitEmptyStrings();
+    private static final MapSplitter TOKEN_SPLITTER =
+            Splitter.on(';').omitEmptyStrings().withKeyValueSeparator('=');
+    private static final CharMatcher QUOTED_STRING_TRIMMER = CharMatcher.is('"');
+
+    @VisibleForTesting
+    static final Function<String, String> FORWARDED_CONVERTER = value -> TOKEN_SPLITTER.split(value).get("for");
+
+    /**
+     * Returns an {@link InetAddress} of a client who initiates a request.
+     *
+     * @param headers the HTTP headers which is received from the client
+     * @param candidateHeaderNames a list of the HTTP header names that may contain a client address
+     * @param proxiedAddresses source and destination addresses retrieved from PROXY protocol header
+     * @param remoteAddress a remote endpoint of a channel
+     * @param filter the filter which evaluates an {@link InetAddress} can be used as a client address
+     * @see <a href="https://tools.ietf.org/html/rfc7239">Forwarded HTTP Extension</a>
+     * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For">X-Forwarded-For</a>
+     */
+    static InetAddress determineClientAddress(HttpHeaders headers,
+                                              List<AsciiString> candidateHeaderNames,
+                                              @Nullable ProxiedAddresses proxiedAddresses,
+                                              InetSocketAddress remoteAddress,
+                                              Predicate<InetAddress> filter) {
+        for (final AsciiString name : candidateHeaderNames) {
+            final InetAddress addr;
+            if (name.equals(HttpHeaderNames.FORWARDED)) {
+                addr = getFirstValidAddress(
+                        headers.get(HttpHeaderNames.FORWARDED), FORWARDED_CONVERTER, filter);
+            } else {
+                addr = getFirstValidAddress(headers.get(name), Function.identity(), filter);
+            }
+            if (addr != null) {
+                return addr;
+            }
+        }
+
+        if (proxiedAddresses != null) {
+            final InetSocketAddress source = proxiedAddresses.sourceAddress();
+            final InetAddress addr = tryResolveInetAddress(source, false);
+            if (addr != null) {
+                return addr;
+            }
+
+            logger.debug("Failed to resolve hostname: {}", source.getHostName());
+        }
+
+        // Return the remote address as unresolved. A user may handle it later.
+        final InetAddress addr = tryResolveInetAddress(remoteAddress, true);
```

Comment:
`Channel.remoteAddress()` is also always resolved. No need to try to resolve here.